### PR TITLE
Rework the lifecycle of singleton instances

### LINKED ISSUE #3817: [BUG] SKFontManager.Default throws an exception
### Description

Calling `SKFontManager.Default` in the latest `4.147.0-preview.1.1` throws an exception

```
Unhandled exception. System.TypeInitializationException: The type initializer for 'SkiaSharp.SKFontManager' threw an exception.
 ---> System.TypeInitializationException: The type initializer for 'SkiaSharp.SKObject' threw an exception.
 ---> System.TypeInitializationException: The type initializer for 'SkiaSharp.SKTypeface' threw an exception.
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at SkiaSharp.SKTypeface..cctor()
   --- End of inner exception stack trace ---
   at SkiaSharp.SKTypeface.EnsureStaticInstanceAreInitialized()
   at SkiaSharp.SKObject..cctor()
   --- End of inner exception stack trace ---
   at SkiaSharp.SKObject..ctor(IntPtr handle, Boolean owns)
   at SkiaSharp.SKFontManager..ctor(IntPtr handle, Boolean owns)
   at SkiaSharp.SKFontManager.SKFontManagerStatic..ctor(IntPtr x)
   at SkiaSharp.SKFontManager..cctor()
   --- End of inner exception stack trace ---
   at SkiaSharp.SKFontManager.get_Default()

```

Tested with both Win32 native assets package and macos native assets package. Happens on .NET 8 and .NET 10

### Code

```cs
using SkiaSharp;
var def = SKFontManager.Default;
```

### Expected Behavior

_No response_

### Actual Behavior

_No response_

### Version of SkiaSharp

Other (Please indicate in the description)

### Last Known Good Version of SkiaSharp

3.116.0 (Current)

### IDE / Editor

Other (Please indicate in the description)

### Platform / Operating System

Windows

### Platform / Operating System Version

- macOS
- Windows 11

### Devices

_No response_

### Relevant Screenshots

_No response_

### Relevant Log Output

```shell

```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ramezgerges**
**Description of Change**

**Bugs Fixed**

- Fixes https://github.com/mono/SkiaSharp/issues/3817

**API Changes**

None.

**Behavioral Changes**

None.

**Required skia PR**

None.

**PR Checklist**

- [ ] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [ ] Merged related skia PRs
- [ ] Changes adhere to coding standard
- [ ] Updated documentation

### CODE REVIEW

**Comment by mattleibow** [comment]
Really nice work on this @ramezgerges — the new `disposeProtected` flow plus making `IgnorePublicDispose` a one-way latch is a much cleaner mental model than the old `SKxxxStatic` subclass dance, and the inline comments explaining the race reasoning are genuinely a pleasure to read 🙌

I took a thorough pass and have a few things to flag before merge. None of these change the overall design — they're mostly small fixes plus some test-coverage suggestions.

### 🔴 Build is currently broken

`binding/SkiaSharp/SKPaint.cs:59` has a typo — `SKFonxt.DefaultSize` should be `SKFont.DefaultSize`. Confirmed locally:

```
SKPaint.cs(59,8): error CS0103: The name 'SKFonxt' does not exist in the current context
```

The new `DefaultFont` lazy path doesn't have direct coverage, which is why CI didn't catch it earlier. A tiny smoke test like `using var p = new SKPaint();` (or exercising `Reset()`) would lock this in.

### 🔴 Likely regression in `SKTypeface.CreateDefault()`

```csharp
return matched == IntPtr.Zero
    ? empty           // private field, null until Empty property is first read
    : GetObject (matched);
```

Pre-refactor `empty` was eagerly initialized in the cctor, so this was always non-null. Now it's lazy, so calling `CreateDefault()` before anyone has touched `SKTypeface.Empty` will return `null`. I think you just want the `Empty` property here.

### 🟠 Racy `GetObject` → `PreventPublicDisposal` pattern still in a few spots

`SKFontManager.MatchFamily` (line 73-74), `MatchCharacter` (177-178), and `SKFontStyleSet` (42, 53) still do:

```csharp
var tf = SKTypeface.GetObject(...);
tf?.PreventPublicDisposal();
```

This is exactly the window the PR closes for singletons: between `GetObject` returning and `PreventPublicDisposal` taking the write lock, the wrapper is HD-observable in an unprotected state, and a concurrent `Dispose()` from another thread that fetched the same cached handle can win. Pre-existing, not something you introduced — but the new `GetDisposeProtectedObject` helper is the perfect tool to migrate these to while you're in the neighbourhood.

### 🟠 Undocumented behavioral change

The PR body says "Behavioral Changes: None", but typefaces returned from `MatchFamily` / `MatchCharacter` are now `IgnorePublicDispose = true`, so user `tf.Dispose()` becomes a silent no-op (cleanup only via finalizer). Worth a one-liner in the PR description / release notes for downstream consumers who deterministically dispose typefaces — they may notice increased native memory pressure.

### 🟡 Perf thought: `Dispose()` now takes the global write lock

The new `Dispose()` holds the HD **write** lock around the `IgnorePublicDispose` check + `isDisposed` CAS. That globally serializes every public Dispose against every reader of the dictionary. For apps that churn lots of short-lived `SKPaint`/`SKPath`/`SKRect` on parallel threads this could become a contention hotspot. A cheap fast-path (volatile read of the flag, CAS, then re-check under the lock only if needed) would let the common non-protected case avoid the writer entirely. Not a blocker — would just be good to benchmark before/after on a busy workload.

### 🟡 `LockRecursionPolicy.SupportsRecursion` — does anything actually recurse?

The PlatformLock comment says it's so Dispose's outer lock can re-enter `DeregisterHandle` via the Handle setter, but in the refactored flow the actual `Dispose(bool)` body runs *outside* the held write lock, so I don't see the recursive path anymore. The upgradeable→write transition inside `GetOrAddObject` → `PreventPublicDisposal` is a legal upgrade, not recursion. Could you double-check whether `SupportsRecursion` is still needed? Dropping back to `NoRecursion` would surface any unintended reentrancy as exceptions in tests.

### 🟡 Test coverage gaps worth filling

Net-new test coverage feels thin given the size of the refactor — and the unreliable repro from #3817 was dropped in `1dd6cfde`. Suggestions in rough priority:

- **Deterministic concurrent-init test** for at least one singleton accessor (`SKColorSpace.CreateSrgb`, `SKTypeface.Default`, `SKFontManager.Default`, `SKData.Empty`, `SKColorFilter.CreateSrgbToLinearGamma`, the `SKBlender` static dict). Two threads barriered on a `ManualResetEventSlim` racing into the accessor, asserting `Assert.Same` and `IgnorePublicDispose == true`. That gets you a real regression net for the issue you're fixing.
- **`SKPaint` construction smoke test** — would have caught the typo above.
- **`SKTypeface.CreateDefault()` when matched == Zero** returns `Empty` — would catch the field-vs-property bug.
- **Concurrent `Dispose` vs `PreventPublicDisposal`** — the central safety claim of the new design, currently only asserted in prose.
- **`GetInstanceNoLocks` filters `IsDisposed` wrappers** — new behavior at `HandleDictionary.cs:139`.
- **`RegisterHandle` replaces a `IsDisposed=true` entry** — new code path.
- **Singleton init under nested factory** — `SKTypeface.Default`'s factory transitively triggers `SKFontManager.Default` and `SKFontStyle.Normal`; worth a test that exercises that chain from a cold start (which is essentially #3817 reproed deterministically).

### ✅ Things I really liked

- The `disposeProtected` flag on `GetOrAddObject` pairing the flag set with the HD critical section is a very nice abstraction.
- One-way-latch `IgnorePublicDispose` with a private setter — exactly right.
- The `LazyInitializer` + per-singleton lock pattern cleanly breaks the cctor cycle without resurrecting the old `EnsureStaticInstanceAreInitialized` hack.
- Eagerly cctor-initing `SKFontStyle` (which has no transitive singleton deps) — correct shortcut.
- `TransferOwnershipToNative` rename + `OwnsHandle = false` *before* `DisposeInternal` closes a real ownership-handoff race, and the stream tests assert the new contract well.
- Scoping `[ModuleInitializer]` to just the native-version check is the minimal correct surface.
- The inline race-reasoning comments in `SKObject.cs` and `HandleDictionary.cs` are excellent — please keep them.

Thanks again for tackling this — it's a tricky area and the refactor is heading somewhere much better than where we were. Happy to chat about any of the above if it'd help. 🙏

**Comment by mattleibow** [comment]
Pushed a follow-up commit (`b3f4041`) on top of your work — wanted to explain the *why* so nothing here is surprising. This is all additive hardening/clarity around the lifecycle you reworked; no behavioural change to the locking design itself.

**Library (`SKObject` / `HandleDictionary`)**
- **Mirror-guard leak fix in `Dispose()`** — the diagnostic re-check (`THROW_OBJECT_EXCEPTIONS`) was throwing *after* claiming disposal but *before* `Dispose(true)`, which would leak the native object in diagnostic builds. It now captures the raced state, runs `Dispose(true)` + `SuppressFinalize`, and only then throws. Production builds are unaffected (guard is compiled out).
- **Comment accuracy** — a few comments overstated the invariant (e.g. implying that observing `IsDisposed` could *only* mean a lock-contract violation, which isn't true given the no-lock internal/finalizer dispose paths your own comment documents). Rewrote them to state the precise invariant, and corrected stale wording on `isDisposed`, `IgnorePublicDispose`, the `disposeProtected` branch, and the `GetOrAddObject` XML doc.
- **Readability** — defined "compare-and-swap (CAS)" on first use and spelled out the `HD` shorthand as `HandleDictionary`.

**Tests**
- Added happy-path + disposed-path coverage for `PreventPublicDisposal`, and a real production-path test that exercises `GetOrAddDisposeProtectedObject` against a *live existing* instance (asserts same-instance return, `IgnorePublicDispose`, not disposed).
- Added `SKSingletonConcurrencyTest` with a **deterministic** re-entrant "GetObject during dispose" interleaving test (forces the disposed-but-not-yet-deregistered window and asserts the disposed wrapper is filtered, replaced without a double-dispose, and the trailing deregister is a no-op), plus a multi-lock singleton-contention canary across all singleton accessors.

Verified locally: build clean, filtered suite **31 passed / 0 failed / 0 skipped** (Debug, with `THROW_OBJECT_EXCEPTIONS` on).

Separately, I'm drafting a *non-blocking* follow-up issue about reducing the global-lock contention via lock striping/sharding (with a benchmark plan) — that's purely an optimisation for later and won't hold up this PR. Really nice work on this rework. 🙏

**Comment by mattleibow** [comment]
Separately tracking some exploratory work on lock-striping / sharding the handle registry over in #4101 (with a draft prototype in #4102) so we don't clutter this PR with it. That investigation is independent and non-blocking — this PR stands on its own.

**Comment by mattleibow** [comment]
I've been doing a deep review and found one regression worth flagging before merge. 

## TL;DR

`SKCodec.Create(Stream)` (and the latent equivalent in `SKFontManager.CreateTypeface(Stream)`) now **crashes the process** with a `NullReferenceException` when the codec actually decodes from a managed `.NET` stream. The new test reproduces it deterministically.

## Reproduction & evidence

The test just decodes after `Create()` (the existing `CanReadManagedStream` only asserts the codec is non-null and never decodes, which is why it didn't catch this):

```csharp
using var stream = new MemoryStream(File.ReadAllBytes(Path.Combine(PathToImages, "baboon.png")));
using var codec = SKCodec.Create(stream);
Assert.NotNull(codec);
Assert.Equal(SKCodecResult.Success, codec.GetPixels(out var pixels)); // 💥 crashes here
```

Running it aborts the test host:

```
Test host process crashed : Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at SkiaSharp.SKManagedStream.OnReadManagedStream(IntPtr buffer, IntPtr size) in SKManagedStream.cs:line 83
   at SkiaSharp.SKManagedStream.OnRead(IntPtr buffer, IntPtr size) in SKManagedStream.cs:line 100
   at SkiaSharp.DelegateProxies.SKManagedStreamReadProxyImplementation(...) in DelegateProxies.stream.cs:line 79
```

## Root cause

This call site changed from `RevokeOwnership(codec)` to `TransferOwnershipToNative()`:

```csharp
// SKCodec.cs:260-261
var codec = GetObject (SkiaApi.sk_codec_new_from_stream (stream.Handle, r));
stream.TransferOwnershipToNative ();   // was: stream.RevokeOwnership (codec);
```

The two are **not** equivalent for managed-callback streams. On `main`, `RevokeOwnership` had two paths:

```csharp
internal void RevokeOwnership (SKObject newOwner)
{
    OwnsHandle = false;
    IgnorePublicDispose = true;
    if (newOwner == null)
        DisposeInternal ();                       // dispose now
    else
        newOwner.OwnedObjects[Handle] = this;     // keep alive, dispose with the owner
}
```

For the codec, `newOwner != null`, so the managed `SKManagedStream` wrapper was **re-parented into `codec.OwnedObjects`** and kept alive until the codec was disposed.

`TransferOwnershipToNative()` always takes the dispose-now path:

```csharp
internal void TransferOwnershipToNative ()
{
    OwnsHandle = false;
    DisposeInternal ();   // always disposes immediately
}
```

Because `WrapManagedStream` builds the wrapper with `disposeManagedStream: true`, `DisposeManaged()` then closes and nulls the underlying `.NET` stream:

```csharp
// SKManagedStream.cs:65-67
if (disposeStream && stream != null) {
    stream.Dispose ();
    stream = null;
}
```

But the native codec reads **lazily** — `sk_codec_new_from_stream` only stores a pointer; the actual reads happen later (e.g. in `GetPixels`) via the managed read callback `OnReadManagedStream`, which now dereferences a null `stream` → NRE across the native→managed boundary → host crash.

### Why it doesn't affect every stream

This only bites streams that native reads back through **managed callbacks** (`SKManagedStream` / `SKFrontBufferedManagedStream`). Native-byte-backed streams like `SKMemoryStream` are fine, because their bytes live in native memory and don't depend on the managed wrapper surviving — which is why the `SKMemoryStream`-based tests stay green and this slipped through.

## Affected call sites

- `binding/SkiaSharp/SKCodec.cs:261` — confirmed crash (test attached).
- `binding/SkiaSharp/SKFontManager.cs` (`CreateTypeface(Stream)` → `TransferOwnershipToNative()`) — same shape. It currently dodges the crash only because it copies into a memory stream first, but it relies on the same eager-dispose semantics and is worth a hard look / matching test.

## Suggested direction (your call)

Restore the "keep the managed wrapper alive until its native owner is disposed" semantics for the callback-stream case — e.g. give the handoff an optional `newOwner` and re-parent into `OwnedObjects` (mirroring `main`'s non-null `RevokeOwnership` path) instead of disposing eagerly, while keeping the always-dispose behavior for native-backed streams. The important invariant: **`Create(Stream)` must not close the caller's stream**, since the codec may decode long after `Create()` returns.

**Comment by mattleibow** [comment]
I added test coverage on top of this — tests only, no production changes, and the latest `main` merged in.

### Corner-case tests for the singleton/lifecycle rework (`SKHandleDictionary*` + helpers)
These drive `HandleDictionary` directly with a fake, non-owning `SKObject` (no native allocations):
- Concurrent construction: callers racing on the same handle get exactly one wrapper, waiters only ever observe a fully-constructed wrapper, and a throwing factory leaves nothing registered so a later call can reconstruct.
- Dispose-protected promotion: `disposeProtected: true` sets the one-way `IgnorePublicDispose` latch, and promote racing a concurrent public `Dispose()` on the same handle always returns a live, protected wrapper (never torn).
- Stale-wrapper / address reuse (ABA): a freed handle builds a fresh wrapper, and a stale wrapper finishing cleanup outside the lock never evicts the live replacement. These assert reference identity (`AssertDeregistered`) rather than "address absent", which is racy under parallel runs.

### Managed-stream ownership handoff (`SKManagedStreamTest`, `SKCodecTest`, `SKTypefaceTest`)
When a .NET `Stream` is passed to `SKCodec.Create(Stream)` or `SKTypeface.FromStream(Stream)`, SkiaSharp wraps it and reads from it lazily (later, on demand), so the wrapper has to outlive the call. These tests lock down that behaviour so the stream-callback fix can't regress:
- The stream stays readable after the call and decoding succeeds — including non-seekable streams, and the lazy SVG/PDF write paths (`CanDecodeManagedStreamAfterCreate`, `...StaysOpenUntilOwnerDisposed`).
- On a failed call (invalid stream) the wrapper is disposed and deregistered immediately instead of leaking.
- If the stream is disposed while the native side is still reading, or both race at teardown, it is closed exactly once with no double-free (`...ClosesManagedStreamExactlyOnce`, `ConcurrentDisposeOf...`).

### Re-entrancy contract (`SKHandleDictionaryReentrancyTest`)
Locks down a factory running inside the upgradeable-read lock: a re-entrant *read* (`GetInstance`) is allowed, while a nested *create* (`GetOrAddObject`) throws `LockRecursionException` on non-Windows. The Windows branch (recursive `CRITICAL_SECTION`, #1383) is skipped since it can't be exercised here.

### Singleton initialization
Two pieces:
- `SKSingletonInitTest` (main test project): each documented singleton — `SKColorSpace.CreateSrgb()`/`CreateSrgbLinear()`, `SKData.Empty`, `SKFontManager.Default`, `SKTypeface.Default`/`Empty`, `SKBlender.CreateBlendMode`, the `SKFontStyle` statics — returns the same instance and carries the `IgnorePublicDispose` flag; `Dispose()` on a protected singleton is a no-op; concurrent access returns one instance; `CreateDefault()` and `new SKPaint()` don't throw or return null. The #3817 cold-start cctor cycle is reproduced via an isolated `AssemblyLoadContext`.
- A separate single-test assembly (`SkiaSharp.Tests.SingletonInit.Console`, wired into cake + sln) whose only test touches `SKFontManager.Default` as the first SkiaSharp type in a fresh `dotnet test` host process — the genuine process-level #3817 regression guard. Keeping it to one test per assembly is what preserves the "first type touched" ordering.

These are here to keep the lifecycle behaviour pinned so a future refactor of this central code can't silently regress it.

File: tests/Tests/SkiaSharp/SKSingletonConcurrencyTest.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+					{
+						errors.Add(ex);
+					}
+				});
+			}
+
+			foreach (var thread in threads)
+				thread.Start();
+			foreach (var thread in threads)
+				thread.Join();
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/SkiaSharp/SKSingletonConcurrencyTest.cs]: Good catch in principle, but I think the current code already covers this. The concurrency test no longer joins bare threads — it delegates to `SKHandleDictionaryTestHelpers.RunConcurrent`, which joins every worker against a shared 30s deadline (`t.Join(remaining)`) and then does `Assert.True(allJoined, deadlockMessage)`. So a genuine product-side deadlock surfaces as a deterministic test failure, not a hung suite/CI run. (See SKHandleDictionaryTestHelpers.cs around the `Stopwatch`/`allJoined` block.)

File: tests/Tests/SkiaSharp/SKSingletonConcurrencyTest.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+						fontManagers[i] = SKFontManager.Default;
+						typefaces[i] = SKTypeface.Default;
+						empties[i] = SKTypeface.Empty;
+						blenders[i] = SKBlender.CreateBlendMode(SKBlendMode.SrcOver);
+					}
+					catch (Exception ex)
+					{
+						errors.Add(ex);
+					}
+				});
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/SkiaSharp/SKSingletonConcurrencyTest.cs]: Agreed on the reasoning, and it's already handled now: every worker in `RunConcurrent` (and the single-thread `RunWithTimeout` helper) is created with `{ IsBackground = true }`, so a timed-out or failed run can't keep the test host alive. The threads were reworked into that shared helper since this review was taken, which is why it isn't visible on this hunk.

File: tests/Tests/SkiaSharp/SKSingletonInitTest.cs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+				object value;
+				try
+				{
+					value = defaultProp.GetValue(null);
+				}
+				catch (TargetInvocationException ex)
+				{
+					throw ex.InnerException;
+				}
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/SkiaSharp/SKSingletonInitTest.cs:220]: Thanks — this is exactly what the current code does. The cold-start test now uses `ExceptionDispatchInfo.Capture(ex.InnerException ?? ex).Throw();` with a trailing `throw;` purely as an unreachable fallback to satisfy definite-assignment. So the original stack trace is preserved and the null-`InnerException` case is handled. Appreciate the suggestion; it matched the fix well.

File: tests/Tests/GarbageCleanupFixture.cs [hunk 4]
```diff
@@ -74,7 +63,9 @@ private bool IsExpectedToBeDead(object instance)
 
 			var skobject = Assert.IsAssignableFrom<SKObject>(instance);
 
-			if (StaticTypes.Contains(skobject.GetType().FullName))
+			// Dispose-protected singleton wrappers are marked with IgnorePublicDispose = true
+			// and live for the process lifetime.
+			if (skobject.IgnorePublicDispose)
 				return false;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/GarbageCleanupFixture.cs:69]: This one is a fair and accurate observation — I dug into it and the concern is real. The old fixture exempted a fixed list of eight nested `*Static` subclasses (`SKData+SKDataStatic`, `SKTypeface+SKTypefaceStatic`, etc.). Those subclasses no longer exist in this PR, so `IgnorePublicDispose` is now the only available signal, and it is strictly broader than that list. I confirmed it is set on genuinely mortal wrappers, not just immortal singletons:

- `SKFontManager.MatchFamily` / `MatchCharacter` and `SKFontStyleSet.CreateTypeface` / `MatchStyle` all return `SKTypeface.GetDisposeProtectedObject(...)`.
- `SKObject.RevokeOwnership(newOwner)` sets it on re-parented managed streams.

So the exemption is wider than before. Two caveats from tracing it through, though:

- Re-parented streams are only reachable via their `newOwner` (added to `OwnedObjects`), and the owner itself is *not* dispose-protected, so an owner leak is still caught by the alive assertion — that sub-case is largely covered indirectly.
- The font-manager / style-set typefaces are the genuine gap: they have no static-field GC root, they are expected to be finalized once unreferenced, and nothing else in the suite would flag them if they leaked.

It is a modest reduction in coverage rather than a correctness bug, but you are right that it dilutes the regression net. If you want to keep the tighter net, one option is a separate, narrower marker (e.g. an `IsImmortalSingleton` flag set only by the static-field accessors, not by the `RevokeOwnership` / font-manager / style-set paths) and exempt on *that* instead of `IgnorePublicDispose`. Happy to hear which way you would prefer.

**Comment by mattleibow** [comment]
## netfx CI root-cause analysis (for the record)

Sharing what the CI logs prove about the recurring `tests-netfx` (net48) redness, so it's documented. TL;DR: the PR is sound and **fixes #3817**, but it surfaces a **pre-existing native heap-corruption flake** on the netfx leg and amplifies it from "retry-recoverable" to "retry-proof." Details below, with PROVEN vs HYPOTHESIS kept strictly separate.

### Proven from CI logs

**1. The netfx failure is native heap corruption, not a failing assertion.**
Every occurrence is `System.AccessViolationException: Attempted to read or write protected memory` wrapped in `TargetInvocationException`, with managed `Failed: 0` (the process is killed). The **victim test is different every time** — e.g. on `main` it landed on `ApiTest.DelegateTypesHaveAttributes`; on this PR's build 158052 it landed on `SKBlenderTest+Gpu`, then `SKPaintTest`, then `SKFrontBufferedManagedStreamTest`. A random innocent victim each time is the fingerprint of heap corruption: the crash lands wherever the next allocation/free happens, not on the guilty code.

**2. The crash pre-exists on `main`.** The exact AV signature appears in **7 of the last 12 `main` netfx logs**, including builds whose netfx leg *succeeded* (retry-masked):

| main build | AV occurrences in log | netfx leg result |
|---|---|---|
| 157962 | 2 | failed (retries exhausted) |
| 157993 / 157996 / 157999 / 158014 / 158024 | 1 each | succeeded (single hit, retry recovered) |
| 158015 | 0 | failed (a different cause) |
| others | 0 | succeeded / canceled |

So on `main`, a single AV per run is absorbed by the run-level retry → green. It only sinks `main` when it recurs (157962).

**3. This PR turns it deterministic + retry-proof.** Of the PR builds that actually compiled and ran netfx, **all of them AV'd**:

| | netfx ran | netfx AV'd | leg green |
|---|---|---|---|
| `main` (12 builds) | 12 | 7 | 9 (retry recovers) |
| this PR (compiled builds: 158030, 158031, 158047, 158048, 158052) | 5 | **5** | **0** |

The two PR builds that got an AzDO job-level retry (158048, 158052) **failed on retry too**; 158052 shows the AV firing 3 separate times, defeating every retry. If the PR behaved like `main` (one recoverable hit/run), P(5/5 runs fail and survive retries) ≈ 0.25⁵ ≈ 0.1%. This is a decisive, statistically significant elevation — not the same flake at the same rate.

**4. The PR fixes #3817.** Worth noting `main` is *also* red right now, but for a different reason: the original #3817 cctor-cycle `NullReferenceException` in `SKTypeface..cctor` on the netcore-Windows leg (build 158029, which is this PR's merge-base). This PR's netcore-Windows leg is green — the lazy-init rework breaks that cycle as intended. 👍

### Still hypothesis (not proven)

*Which* change amplifies the corruption. Two candidates I can't separate from logs alone:
- **(A) Production** — the lazy-singleton + dispose-protection refcount reconciliation path (`GetOrAddObject(... unrefExisting / disposeProtected ...)`, `IgnorePublicDispose`) double-frees / uses-after-free a native singleton more often.
- **(B) New stress tests** — the added concurrency tests (`SKSingletonConcurrencyTest`, `SKManagedStreamConcurrencyTest`, `HandleDictionaryThreadingCollection`) hammer the machinery and merely *expose* the latent pre-existing race far more frequently.

### Suggested next steps to settle (A) vs (B)
1. **Local ASAN repro** — build native Skia with AddressSanitizer and run the concurrency tests under netcore; this should catch the use-after-free deterministically with a real native stack, independent of net48.
2. **CI isolation** — a throwaway run with the new concurrency stress tests disabled; if netfx goes green, the amplifier is the tests exposing a pre-existing latent bug rather than a regression in the production rework.

Either way the rework itself looks correct and the #3817 fix is solid — this is about taming a flake that was already lurking on `main`. Happy to help drive the ASAN repro if useful. Really nice work on this. 🙏

*(Analysis from CI build logs on xamarin/public def 4; builds 157962–158052.)*

**Comment by mattleibow** [comment]
Hey @ramezgerges 👋 — updating my earlier comment now that I've stopped reasoning-by-elimination and actually **instrumented and reproduced** this. The good news: the root cause I described holds up, and I can now name the freed objects and show the crash deterministically. The one correction is that it's **not fundamentally netfx-specific** — netfx CI just makes it *visible*.

Thanks again for this rework — the explicit lifecycle/lock comments are what made it possible to instrument this cleanly.

## TL;DR (now proven, not inferred)

The crash is a **use-after-free of a shared process-global Skia object**. Concretely measured victims: the **default `SKTypeface`** and the **sRGB `SKColorSpace`** (the same class applies to `SKData.Empty`, default `SKFontManager`, etc.). Two changes in this PR combine to reopen native-unref paths on those globals that `main` never had:

1. **Eager → lazy init** (removal of `static SKObject()` / `EnsureStaticInstanceAreInitialized()`).
2. **`owns:false` + no-op `Dispose(bool)` → `owns:true` + dispose-protected** (`IgnorePublicDispose = true`).

Individually fine; together they make a shared global reachable by an unref path.

## Why dispose-protected ≠ immortal (the crux)

`IgnorePublicDispose` is checked in **exactly one** place — the public `Dispose()`. There are **three** teardown paths:

| Teardown path | Guarded by `IgnorePublicDispose`? |
|---|---|
| public `Dispose()` | ✅ yes |
| `DisposeInternal()` (ownership handoff, owned-child cleanup, stale-wrapper dict replacement) | ❌ **no** |
| finalizer `~SKNativeObject` → `Dispose(false)` | ❌ **no** |

Both unguarded paths reach `DisposeNative()` → `SafeUnRef()` when `Handle != 0 && OwnsHandle`. Since singletons are now `owns:true`, **dispose-protection does nothing to stop `DisposeInternal()` or the finalizer from unref'ing the shared global.** Your own `SKObject` comment says it: *"the wrapper is NOT immortal from GC's perspective; finalization and DisposeInternal still tear it down."* So `dispose-protected ⊊ immortal`, and singletons shipped here are exactly the dangerous middle case.

## The proof (local A/B, macOS .NET Core)

I added a refcount **ledger** that captures each singleton's exact native refcount, then drives the two reachable over-unref vectors, and ran it with the immortal latch **off** (this PR's config) vs **on** (the fix). Same binary, one-line toggle, native lib unchanged.

Measured baseline: `SKTypeface.Default` refcount = **2**, and `SKTypeface.CreateDefault()` returns the **same native handle** as `SKTypeface.Default` (`sameAsDefault = True`) — so `CreateDefault` is literally a mortal `owns:true` sibling wrapper over the shared default-typeface global.

| Probe | `immortal:false` (this PR) | `immortal:true` (fix) |
|---|---|---|
| `DisposeInternal()` on sRGB singleton | 💥 **host process crash** (heap corruption) | ✅ no-op, refcount stable |
| `DisposeInternal()` on default typeface | 💥 **host process crash** | ✅ no-op, refcount stable |
| `CreateDefault()` transient churn + GC | ❌ refcount falls **below baseline** (named over-unref) | ✅ stays ≥ baseline |
| sRGB churn control (distinct color spaces) | ✅ pass | ✅ pass |
| Full singleton suite (22 tests) | crashes / asserts fail | ✅ **22/22 green** |

So there are **two** independently reachable vectors, both now reproduced:

1. **`DisposeInternal()` on a singleton** — the cleanest, deterministic, *platform-independent* one. It crashes the test host **on macOS .NET Core**, which is why I can now say this isn't a netfx quirk: it's plain heap corruption. The immortal latch turns it into a no-op.
2. **Transient sibling wrappers** (`CreateDefault()`, and any `owns:true` transient minted over a shared handle during the lazy window) — their finalizer/`DisposeInternal` runs `SafeUnRef()` on the shared global and drops it below baseline. This one is GC/finalizer-timing-sensitive, which is exactly why it surfaces *reliably as an AVE on the netfx CI leg* (deterministic finalizer flush + corrupted-state-exception escalation) while .NET Core mostly tolerates the dip — but as vector (1) shows, the underlying corruption is the same everywhere.

The original "netfx-specific because of finalizer timing" framing was therefore half-right: the *observation* is netfx-biased, the *bug* is not.

## What `main` did (the guarantee to restore)

`main` gave singletons **three stacked guarantees**: rooted in a `static readonly` field (finalizer never runs), `owns:false` (no `DisposeNative`), and a no-op `Dispose(bool)` (zero unref paths) — via the `SKTypefaceStatic`-style subclasses, eagerly seeded in `static SKObject()`. Exactly one immortal wrapper per global, from first touch. This PR collapses that onto the single weaker `IgnorePublicDispose` concept.

## Fix

Re-introduce a true **immortal latch**, generalized so we don't need the old `*Static` subclasses:

- One-way `isImmortalSingleton` latch via `GetOrAddImmortalSingletonObject(...)` (routes singletons through `HandleDictionary` with `immortal: true`), which also **promotes any pre-existing dict entry** for that handle to immortal — closing the lazy-window race where a transient registered first.
- The latch is checked at the **top of all three teardown entry points** — finalizer, `DisposeInternal()`, and public `Dispose()` — **before** the `isDisposed` CAS, so no path can unref the shared global.

This restores `main`'s unconditional immortality for singletons and leaves `owns:true` semantics untouched for everything else. A/B above: 22/22 green with the latch; crashes/over-unrefs without it.

## CI corroboration

Consistent with the local A/B: the netfx leg AVE'd 3× on this branch's config (top frame `AnimationBuilderTest.CanRenderWithBase64` — a corruption *victim*, not the culprit), and went green across 4/4 confirmation builds with the latch.

I've been iterating the latch + the ledger test on a scratch branch so I don't churn your PR. Happy to push it here as a commit or hand it over as a patch — and I'd suggest keeping the two `DisposeInternal*` ledger probes as a permanent regression guard, since they fail *loudly* (host crash) the moment immortality regresses. Genuinely nice work on this PR — this is just restoring one guarantee the cleanup relaxed. 🙏
